### PR TITLE
Producer reconnect to broker

### PR DESCRIPTION
**Pull request by dvladi77**

### CODE REVIEW

File: nmsg/kafkaio.c [hunk 1]
```diff
@@ -489,13 +512,52 @@ _kafka_delivery_cb(rd_kafka_t *rk, const rd_kafka_message_t *rkmessage, void *op
...
 	ctx->delivered++;
 }
 
+static void
+_kafka_rebalance_cb(rd_kafka_t *rk, rd_kafka_resp_err_t err, rd_kafka_topic_partition_list_t *partitions, void *opaque)
+{
+	rd_kafka_error_t *error = NULL;
+	rd_kafka_resp_err_t ret_err = RD_KAFKA_RESP_ERR_NO_ERROR;
+	kafka_ctx_t ctx = (kafka_ctx_t) opaque;
+	if (ctx == NULL)
```
Comment (djw1149) [inline nmsg/kafkaio.c:529]: Should this be noisier if ctx is NULL?

File: nmsg/kafkaio.c [hunk 2]
```diff
@@ -489,13 +512,52 @@ _kafka_delivery_cb(rd_kafka_t *rk, const rd_kafka_message_t *rkmessage, void *op
...
 	}
+	ctx->error_retry = 0;
 	ctx->delivered++;
 }
 
+static void
+_kafka_rebalance_cb(rd_kafka_t *rk, rd_kafka_resp_err_t err, rd_kafka_topic_partition_list_t *partitions, void *opaque)
+{
+	rd_kafka_error_t *error = NULL;
+	rd_kafka_resp_err_t ret_err = RD_KAFKA_RESP_ERR_NO_ERROR;
```
Comment (djw1149) [inline nmsg/kafkaio.c:527]: suggest s/ret_err/resp_err/

It seems gross that kafka has two types of "error".

File: nmsg/kafkaio.c [hunk 3]
```diff
@@ -489,13 +512,52 @@ _kafka_delivery_cb(rd_kafka_t *rk, const rd_kafka_message_t *rkmessage, void *op
...
+		_nmsg_dprintf(3, "%s: partitions revoked (%s):\n", __func__, rd_kafka_rebalance_protocol(rk));
+
+		if (!strcmp(rd_kafka_rebalance_protocol(rk), "COOPERATIVE"))
+			error = rd_kafka_incremental_unassign(rk, partitions);
+		else
+			ret_err = rd_kafka_assign(rk, NULL);
+		break;
+        default:
+		_nmsg_dprintf(2, "%s: failed: %s\n", __func__, rd_kafka_err2str(err));
+		rd_kafka_assign(rk, NULL);
```
Comment (djw1149) [inline nmsg/kafkaio.c:550]: should you assign the result to ret_err and check it?

File: nmsg/kafkaio.c [hunk 4]
```diff
@@ -470,13 +484,22 @@ _kafka_error_cb(rd_kafka_t *rk, int err, const char *reason, void *opaque)
 	rd_kafka_resp_err_t err_kafka = (rd_kafka_resp_err_t) err;
 
 	switch(err_kafka) {
```
Comment (djw1149) [inline nmsg/kafkaio.c:486]: add space after "switch"

**Review by shw700 — APPROVED** [review]

**Pull request merged**